jb/itch5/testing/ut_type_based_order_book.hpp

```diff
@@ -0,0 +1,308 @@
+#ifndef jb_itch5_ut_type_based_order_book_hpp
+#define jb_itch5_ut_type_based_order_book_hpp
+
+#include <boost/test/unit_test.hpp>
+
+namespace jb {
+namespace itch5 {
+namespace testing {
+
+/**
+ * side_type class trivial member tests.
+ * @tparam side_type Side type to be tested
+ *
+ * Uses testing hook is_ascending() to know if tested is buy or sell side
+ */
+template <typename side_type>
+void test_side_type_trivial(side_type& tested) {
+  using jb::itch5::price4_t;
+  auto is_less = tested.is_ascending();
+  auto actual = tested.best_quote();
+  if (is_less) {
+    BOOST_CHECK_EQUAL(actual.first, price4_t(0));
+    BOOST_CHECK_EQUAL(actual.second, 0);
+  } else {
+    BOOST_CHECK_EQUAL(actual.first, price4_t(200000UL * 10000));
+    BOOST_CHECK_EQUAL(actual.second, 0);
+  }
+
+  actual = tested.worst_quote();
+  if (is_less) {
+    BOOST_CHECK_EQUAL(actual.first, price4_t(0));
+    BOOST_CHECK_EQUAL(actual.second, 0);
+  } else {
+    BOOST_CHECK_EQUAL(actual.first, price4_t(200000UL * 10000));
+    BOOST_CHECK_EQUAL(actual.second, 0);
+  }
+
+  //  book_depth should be 0
+  BOOST_CHECK_EQUAL(tested.count(), 0);
+}
+
+/**
+ * Test side type error handling.
+ * @tparam side_type Side to be tested
+ */
+template <typename side_type>
+void test_side_type_errors(side_type& tested) {
+  using jb::itch5::price4_t;
+
+  int diff;
+  if (tested.is_ascending()) {
+    diff = -10000;
+  } else {
+    diff = 10000;
+  }
+
+  // Add two orders to the book ...
+  (void)tested.add_order(price4_t(100000), 100);
+  (void)tested.add_order(price4_t(100000 - diff), 200);
+
+  // ... check the best quote ...
+  auto actual = tested.best_quote();
+  BOOST_CHECK_EQUAL(actual.first, price4_t(100000 - diff));
+  BOOST_CHECK_EQUAL(actual.second, 200);
+
+  // ... remove the first order, once should work, the second time
+  // should fail ...
+  tested.reduce_order(price4_t(100000), 100);
+  BOOST_CHECK_THROW(tested.reduce_order(price4_t(100000), 100), jb::feed_error);
+
+  // ... check the best quote again ...
+  actual = tested.best_quote();
+  BOOST_CHECK_EQUAL(actual.first, price4_t(100000 - diff));
+  BOOST_CHECK_EQUAL(actual.second, 200);
+  // .. and the book_depth should be incremented
+  BOOST_CHECK_EQUAL(tested.count(), 1);
+
+  // ... reduce a non existing price better than the inside
+  BOOST_CHECK_THROW(
+      tested.reduce_order(price4_t(100000 + 2 * diff), 100), jb::feed_error);
+
+  // ... reduce non existing order on an empty bottom levels
+  if (diff < 0) {
```

Comment:
I think you are using diff <  0 and is_ascending interchangeably, please consider using only one of them.